src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
         }
         #endregion
     }
+
+    public readonly struct HttpHeadersNonValidated : IEnumerable<KeyValuePair<string, HeaderStringValues>>
+    {
+        private readonly IEnumerable<KeyValuePair<HeaderDescriptor, object>>? _headerStore;
+
+        internal HttpHeadersNonValidated(IEnumerable<KeyValuePair<HeaderDescriptor, object>>? headerStore)
+        {
+            _headerStore = headerStore;
+        }
+
+        public IEnumerator<KeyValuePair<string, HeaderStringValues>> GetEnumerator() => _headerStore != null
+            ? new Enumerator(_headerStore)
+            : ((IEnumerable<KeyValuePair<string, HeaderStringValues>>)Array.Empty<KeyValuePair<string, HeaderStringValues>>()).GetEnumerator();
+
+        IEnumerator<KeyValuePair<string, HeaderStringValues>> IEnumerable<KeyValuePair<string, HeaderStringValues>>.GetEnumerator() => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<KeyValuePair<string, HeaderStringValues>>
+        {
+            private readonly IEnumerator<KeyValuePair<HeaderDescriptor, object>> _headerStore;
+
+            internal Enumerator(IEnumerable<KeyValuePair<HeaderDescriptor, object>> headerStore)
+            {
+                _headerStore = headerStore.GetEnumerator();
+                Current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (!_headerStore.MoveNext())
+                {
+                    return false;
+                }
+
+                KeyValuePair<HeaderDescriptor, object> nextPair = _headerStore.Current;
+
+                Current = new KeyValuePair<string, HeaderStringValues>(nextPair.Key.Name, ToStringValues(nextPair.Key, nextPair.Value));
+                return true;
+
+                static HeaderStringValues ToStringValues(HeaderDescriptor descriptor, object value)
+                {
+                    if (value is HttpHeaders.HeaderStoreItemInfo info)
+                    {
+                        if (info.OriginalRawValue != null)
+                        {
+                            return info.OriginalRawValue is string rawString ? new HeaderStringValues(rawString) : new HeaderStringValues((List<string>)info.OriginalRawValue);
+                        }
+
+                        return new HeaderStringValues(HttpHeaders.GetValuesAsStrings(descriptor, info));
+                    }
+
+                    Debug.Assert(value is string);
+                    return new HeaderStringValues((string)value);
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
```

Comment:
Just a personal styling choice, but having the public implementation of an interface getter enhanced with setter seems a bit confusing.
Wouldn't it be a slightly readable to back it up with private `_current` field and leave the property to contain only getter and thus match the interface?

Reply:
Personally, I'd prefer to have less code when it's possible and keep the current version. Adding a private doesn't change the interface in any way.